### PR TITLE
Feat: Add NBFix and MBAR estimator

### PR DESCRIPTION
**Pull request by WangXinyan940**
Refactor:
1. Split old api.py to separate files. Now the api.py only contains Hamiltonian class. The generators are moved to dmff/generators/*.py

New Features:
1. LennardJonesForce and NBFix support.
2. MBAR estimator for free energy estimation and reweighting.

Bug fix:
1. Use template name instead of residue name for atomic charge matching.

### CODE REVIEW

File: dmff/mbar.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+from jax import grad
+from tqdm import tqdm, trange
+
+
+class BaseState:
+    def __init__(self, temperature, name):
+        self._temperature = temperature
+        self.name = name
+
+    def calc_energy_frame(self, frame):
```
Comment (KuangYu) [inline dmff/mbar.py:15]: Following the convention of JAX, this function is better to be passed as an argument in the constructor.

File: dmff/mbar.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    def add_state(self, state):
+        self.states.append(state)
+
+    def remove_sample(self, name):
+        init_num = len(self.samples)
+        self.samples = [s for s in self.samples if s.from_state != name]
+        final_num = len(self.samples)
+        assert init_num < final_num
+
+    def remove_state(self, name):
```
Comment (KuangYu) [inline dmff/mbar.py:63]: When a state is removed, it's corresponding samples should be removed too

File: dmff/mbar.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
+import numpy as np
+import mdtraj as md
```
Comment (KuangYu) [inline dmff/mbar.py:2]: mdtraj should be added in requirement.txt

File: dmff/mbar.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            argsort = np.argsort(wnew)[::-1][:int(eff_samples) + 1]
+            for nstate in range(state_effect.shape[0]):
+                istart = self._nk[:nstate].sum()
+                iend = istart + self._nk[nstate]
+                state_effect[nstate] = ((argsort > istart) &
+                                        (argsort < iend)).sum()
+            return eff_samples, state_effect
+        return eff_samples
+
+    def estimate_free_energy(self, unew):
```
Comment (KuangYu) [inline dmff/mbar.py:184]: 1. We should consider computing free energies between two target states.
2. Target states and params should be passed in as argument.

**Review by KuangYu — CHANGES_REQUESTED** [review]
See review

File: dmff/mbar.py (hunk 3 above)
Comment (WangXinyan940) [reply dmff/mbar.py:2]: Done

File: dmff/mbar.py (hunk 1 above)
Comment (WangXinyan940) [reply dmff/mbar.py:15]: For Jax potential, we can use this way. For external MD engines, we'd better create new class for customize initialization of MD engine.
The design is updated in the latest version.

File: dmff/mbar.py (hunk 2 above)
Comment (WangXinyan940) [reply dmff/mbar.py:63]: Done

**Comment by Ericwang6** [comment]
I've removed some unused imports

**Review by KuangYu — CHANGES_REQUESTED** [review]
See comments

**Review by KuangYu — APPROVED** [review]

**Pull request merged**